[share] Add sharing file support (android & ios)

pboos opened the pull request:
This pull request adds file sharing support to the share plugin.

``` dart
Share.shareFile(File('${directory.path}/image.jpg'));
```

To use while it is not yet published in the share plugin, just use the following in your `pubspec.yaml`:
``` yaml
dependencies:
  ...
  share:
    git:
      url: https://github.com/pboos/plugins.git
      path: packages/share
      ref: feature/shareFile
```

palancana commented:
Hi @pboos! What a nice work! Thank you very much for sharing, I find this completely necessary for the plugin!
One quick comment. To make things easier for people who don't know about this, I think we should mention in the README.md that adding some keys for iOS may be needed. For Android I think it's not needed, correct me if I'm wrong please. We could add something like this:

**iOS**
Add these keys to ios > Runner > Info.plist to be able to same images to the camera roll:
```
<key>NSPhotoLibraryUsageDescription</key>
<string>Access to save the image</string>
<key>NSPhotoLibraryAddUsageDescription</key>
<string>Access to save the image</string>
```
**Android**
There's no need to change anything.

allanwolski commented:
Great job @pboos!

It would be perfect if had an option for multiple files like this
https://github.com/d-silveira/flutter-share

Could you add that?

JorgeSardina commented:
I'm having issues sharing image files with WhatsApp on iOS. On Android it's working fine, is there any specific thing that needs to be done in order to share in WhatsApp?

allanwolski commented:
When will this be merged?

kitesoft commented:
When will this be merged?

woellij commented:
using the dependency as described by @pboos does not seem to work anymore with the migration away from android support library

```
E:\Dev\flutter\flutter\.pub-cache\git\plugins-32d70052e89ccb4ea8c528f174dd29023c037a16\packages\share\android\src\main\java\io\flutter\plugins\share\SharePlugin.java:10: error: package android.support.annotation does not exist
import android.support.annotation.NonNull;
                                 ^
E:\Dev\flutter\flutter\.pub-cache\git\plugins-32d70052e89ccb4ea8c528f174dd29023c037a16\packages\share\android\src\main\java\io\flutter\plugins\share\SharePlugin.java:11: error: package android.support.v4.content does not exist
import android.support.v4.content.FileProvider;
                                 ^
E:\Dev\flutter\flutter\.pub-cache\git\plugins-32d70052e89ccb4ea8c528f174dd29023c037a16\packages\share\android\src\main\java\io\flutter\plugins\share\SharePlugin.java:155: error: cannot find symbol
  @NonNull
   ^
  symbol:   class NonNull
  location: class SharePlugin
E:\Dev\flutter\flutter\.pub-cache\git\plugins-32d70052e89ccb4ea8c528f174dd29023c037a16\packages\share\android\src\main\java\io\flutter\plugins\share\SharePlugin.java:101: error: cannot find symbol
        FileProvider.getUriForFile(
        ^
  symbol:   variable FileProvider
  location: class SharePlugin
4 errors

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':share:compileDebugJavaWithJavac'.
> Compilation failed; see the compiler error output for details.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 5s
*******************************************************************************************
The Gradle failure may have been because of AndroidX incompatibilities in this Flutter app.
See https://goo.gl/CP92wY for more information on the problem and how to fix it.
*******************************************************************************************
Finished with error: Gradle task assembleDebug failed with exit code 1
```

yhunglee commented:
@pboos Hi, I'm an user of your great work. Thank you.  I appreciate your great effort. It can be better if you make it to be merged. I wish it become true. Thank you.

BGomes17 commented:
Using this plugin in the way that @pboos described is giving a conflict with AndroidX.
Is there any forecast when this feature will be added to the share plugin?

martinory commented:
I found it easy to refactor the plugin according to https://flutter.dev/docs/development/packages-and-plugins/androidx-compatibility as well as changing the NonNull and FileProvider to the latest imports in the SharePlugin.java-file:
`import androidx.annotation.NonNull;`
`import androidx.core.content.FileProvider;`
and in the androidmanifest.xml-file:
`android:name="androidx.core.content.FileProvider"`

matheuscaldasrj commented:
Can anyone check this fail check?

tomwassing commented:
Is anyone able to fix the CLA related check?

collinjackson commented:
At first glance looks like a solid PR.

Created https://github.com/flutter/flutter/issues/36894 to track getting this merged, if this feature is something you need please thumbs up that issue.

bulwinkel commented:
@collinjackson anything we can do to help this along?

pboos commented:
What do I have to do to get this merged in? It seems like several people are interested in this and it would make it easier if it was merged into the official share plugin.

joandervieira commented:
I also have problem conflict with AndroidX. My project not use AndroidX

`Attribute meta-data#android.support.FILE_PROVIDER_PATHS@resource value=(@xml/filepaths) from [:open_file] AndroidManifest.xml:20:17-50
	is also present at [:share] AndroidManifest.xml:15:17-65 value=(@xml/flutter_share_file_paths).
	Suggestion: add 'tools:replace="android:resource"' to <meta-data> element at AndroidManifest.xml to override.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:processDebugManifest'.
> Manifest merger failed : Attribute meta-data#android.support.FILE_PROVIDER_PATHS@resource value=(@xml/filepaths) from [:open_file] AndroidManifest.xml:20:17-50
  	is also present at [:share] AndroidManifest.xml:15:17-65 value=(@xml/flutter_share_file_paths).
  	Suggestion: add 'tools:replace="android:resource"' to <meta-data> element at AndroidManifest.xml to override.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 19s`

JonasVautherin commented:
@pboos: I tried to fix the conflicts (your repo is 500+ commits behind upstream), but unfortunately I couldn't make the plugin work.

Would you mind fixing the conflicts (it seems to me that it shouldn't be too much work)? Also Flutter is complaining about AndroidX in my case, and it seems like this should be addressed as well.

Finally, the CI is all passing expect for the CLA saying: "CLAs are signed, but unable to verify author consent". Did you do everything on your side? I believe that the Flutter team won't be able to merge unless this test is passing.

It seems like the Flutter team doesn't have much resources to look into PRs, since this one has been waiting for more than a year. But at least if it was in a mergeable state, there would be a chance of it being merged some day :blush:. Currently it is not.

pboos commented:
So. I have updated my branch and rebased to a commit from two weeks ago (as of now the most recent). The many conflicts were due to the reason that the master share plugin was migrated to android plugin v2 and androidx. I now adjusted the code to that version.

Might require a new review of the code since especially the ios part changed a log (made it closer to the new version and as well fixed a bug that was there in the old version of my pull request).

Any further input to get this merged would be great.

JonasVautherin commented:
Happy to see it's moving here! And I see that the CLA test is now passing!

I had a look a the CI: many tests fail for the same reason, but I'm not sure it comes from this PR:

```
Running version check for changed packages
Unhandled exception:
NoSuchMethodError: The getter 'nextMajor' was called on null.
Receiver: null
Tried calling: nextMajor
#0      Object.noSuchMethod (dart:core-patch/object_patch.dart:53:5)
#1      getAllowedNextVersions (package:flutter_plugin_tools/src/version_check_command.dart:59:19)
#2      VersionCheckCommand.run (package:flutter_plugin_tools/src/version_check_command.dart:153:13)
<asynchronous suspension>
#3      CommandRunner.runCommand (package:args/command_runner.dart:197:27)
#4      CommandRunner.run.<anonymous closure> (package:args/command_runner.dart:112:25)
#5      new Future.sync (dart:async/future.dart:224:31)
#6      CommandRunner.run (package:args/command_runner.dart:112:14)
#7      main (package:flutter_plugin_tools/src/main.dart:57:17)
#8      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:305:32)
#9      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:174:12)
```

There is an analysis error (I'll comment in the code) and the formatting is wrong. At least those two I think you can fix :smile:.

Note: I'm new to Flutter, I'm just trying to help :sweat_smile:.

JonasVautherin commented on packages/share/example/lib/main.dart:
```diff
@@ -83,4 +133,11 @@ class DemoAppState extends State<DemoApp> {
     );
   }
+
+  Future<File> _createTextFile(String filename, String content) async {
```
This is failing the [analyze](https://github.com/flutter/plugins/pull/970/checks?check_run_id=372549971) test with:

```Checking project share...

  info - The class 'Future' wasn't exported from 'dart:core' until version 2.1, but this code is required to be able to run on earlier versions at example/lib/main.dart:137:3 - (sdk_version_async_exported_from_core)

1 issue found; analyzed 9 source files in 19.6s.



The following packages have analyzer errors (see above):
 * share```

pboos replied:
I hope with my new commit this should be resolved.

pboos commented:
So formatting and the async should be fixed now. Just the `NoSuchMethodError: The getter 'nextMajor' was called on null.` is still happening now. I wonder where this one is coming from and how I can fix it...

JonasVautherin commented:
Agreed, they seem fixed. The `nextMajor` error seems to come from the tool [`version_check_command.dart`](https://github.com/flutter/plugin_tools/blob/master/lib/src/version_check_command.dart#L59), apparently because the [`masterVersion`](https://github.com/flutter/plugin_tools/blob/master/lib/src/version_check_command.dart#L144-L145) given as an argument is null:

```
final Version masterVersion =
    await gitVersionFinder.getPackageVersion(pubspecPath, baseSha);
```

@collinjackson is that something you could help with? :blush:

JonasVautherin commented:
BTW what's the difference between this and #1423?

pboos commented:
If I read correctly, then the problem is: the MASTER branch in the example/pubspec.yaml hast NO version. Because of that, masterVersion will be null. So there is NO WAY this will pass until the master branch is fixed :D

About #1423:
That build probably succeeded because that change might have been effective after their commit. Change seems to have been done in June 2019. Not sure when it went into master and therefore into the checks...

About the code difference:
That one is based on this pull request. Same code but fixed to work with the new v2 plugin system and androidx. So basically the same as here. Just on iOS there are small differences (i rewrote some part of it this time). There is one small bug on iOS that I think exists on that version that I fixed in my new rewrite.

pboos commented:
How to get this merged in? Not sure if anything else needs to be done or if this just takes this long until it can get reviewed.

JonasVautherin commented:
The usual way, as far as I know, is that at some point we would need to hear about a maintainer. They may have comments on the code, but everything is ready for a review.

Kindly pinging @cyanglaz and @collinjackson, hoping to see a sign of interest at some point :see_no_evil:.

ejain commented:
Just tested this on iPadOS 13.3 (simulator). The dialog showed as expected, but neither _Copy_ nor _Save to Files..._ worked. This may be because I passed in a `MemoryFileSystem` file (how about accepting bytes instead of a file?). Moreover, _Save to Files..._ may require additional permissions (beyond `UIFileSharingEnabled`).

JonasVautherin commented:
> Can't get Save to Files... to work in the simulator, but that's a pre-existing condition...

@ejain: I'm not sure I get that. Does that mean that it is expected behavior, or does it mean that this PR would need to be fixed for that case?

ejain commented:
> @ejain: I'm not sure I get that. Does that mean that it is expected behavior, or does it mean that this PR would need to be fixed for that case?

The _Save to Files_ option doesn't work for me on simulated iPhones, either, even though it does work on the physical iPhone I tested, so I don't think this is an issue with this PR.

JonasVautherin commented:
Got it! Then if you believe it is ready to be merged (or at least reviewed by a maintainer), feel free to "approve" the PR. Makes it easier to track, IMO.

collinjackson commented on packages/share/example/android/app/src/main/java/io/flutter/plugins/shareexample/EmbeddingV1Activity.java:
```diff
@@ -1,18 +0,0 @@
-import io.flutter.app.FlutterActivity;
-import io.flutter.plugins.GeneratedPluginRegistrant;
-
-public class EmbeddingV1Activity extends FlutterActivity {
```
Why was this deleted? If it shouldn't be deleted, can we bring it back?

pboos replied:
Ah true. Guess I deleted too much :) Will bring it back.

pboos replied:
Noticed now, why I switched totally to v2. Reason is, that using v1 didn't work because path_provider seems to only work with v2? Build works fine with v1 example as it is now in the share example. But as soon as we add path_provider (used in the example to store a text file and share it), it will not work with the error:

MissingPluginException(No implementation found for method getApplicationDocumentsDirectory on channel plugins.flutter.io/path_provider)

Any recommendation on what to do?

Sorry for the force pushes. Trying to clean up the commit log with all the small forth and back.

pboos replied:
Never mind. Had to add the path provider plugin manually in the MainActivity.java to have it available.

collinjackson commented on packages/share/lib/share.dart:
```diff
@@ -51,4 +52,93 @@ class Share {
+  /// Summons the platform's share sheet to share a file.
+  ///
+  /// Wraps the platform's native share dialog. Can share a file.
+  /// It uses the ACTION_SEND Intent on Android and UIActivityViewController
```
```suggestion
  /// It uses the `ACTION_SEND` Intent on Android and `UIActivityViewController`
```

collinjackson commented:
I believe that the version should be updated to 0.6.4

pboos commented:
@collinjackson Fixed the suggestions and updated the version. If there is anything else, just let me know.

ejain commented:
Went ahead and submitted an app that includes this change (minus any commits added in the past day or so), but according to the app store reviewer, the file sharing dialog doesn't show on their iPad :-(

ejain commented:
If anyone here has a (physical) iPad for testing: I have a simple app in beta that makes use of the shareFile from this pull request at https://testflight.apple.com/join/2yFSm6KQ. Source code is at https://github.com/zenobase/zenobase-durations/.

pboos commented:
Sharing a file and saving it as a file worked without a problem for me on the emulator. Here the proof:
[Simulator Screen Shot - iPad Pro (12 9-inch) (3rd generation) - 2020-01-16 at 07 45 23](https://user-images.githubusercontent.com/398400/72500349-8dff6000-3834-11ea-8996-6bf405c00a73.png)
[Simulator Screen Shot - iPad Pro (12 9-inch) (3rd generation) - 2020-01-16 at 07 45 54](https://user-images.githubusercontent.com/398400/72500572-239aef80-3835-11ea-990a-ab5fe237ae98.png)
[Simulator Screen Shot - iPad Pro (12 9-inch) (3rd generation) - 2020-01-16 at 07 47 27](https://user-images.githubusercontent.com/398400/72500350-8e97f680-3834-11ea-8371-0943d7a51d5f.png)
[Simulator Screen Shot - iPad Pro (12 9-inch) (3rd generation) - 2020-01-16 at 07 48 04](https://user-images.githubusercontent.com/398400/72500351-8e97f680-3834-11ea-9b07-c2dcd0b67454.png)

What I noticed is, that it saves two files (the shared file and the additional text). If one removes the additional text and only shares file and subject, it only creates one file (see next screenshot)
[Simulator Screen Shot - iPad Pro (12 9-inch) (3rd generation) - 2020-01-16 at 07 55 30](https://user-images.githubusercontent.com/398400/72500799-973cfc80-3835-11ea-85a0-c156902da592.png)

PS: Thinking of updating the example to actually share an image coming from "image_picker". Maybe that example will clearer than sharing a created text file.

pboos commented:
Just noticed, that it could be possible to share multiple files. I wonder if I should update this pull request to add support for it (while still supporting single file sharing as well).

JonasVautherin commented:
> Just noticed, that it could be possible to share multiple files. I wonder if I should update this pull request to add support for it.

I wouldn't. If you keep adding stuff into a PR, it makes it more difficult for maintainers to merge it. Take it like this: they review your PR and ask for changes. If you then add new things, they need to review the PR again. So in some way you cancel the previous review and the process starts over.

I would focus on getting this PR merged, and then you can always open new PRs later if you want to add new features :+1:.

ejain commented:
While this also works for me in the simulator, I just heard from a second tester (with a 6th gen iPad running iOS 13.3) that the share dialog still doesn't show.

JonasVautherin commented:
Dialog shows for me on my iPad, but whatever I choose (save to file, mail, ...) crashes the app, and I don't seem to get a useful error log. Though I did not try with the example, so maybe my code is missing something.

@ejain: did you try with @pboos example code?

pboos commented:
Sadly I do not have an iPad. Only an iPhone for testing. And in the iPad emulator everything went fine.

ejain commented:
_Save as File_ works for me on an iPhone SE with iOS 13.3, but not on any device in the Simulator, so I don't think it's an issue with this pull request. But as mentioned above, I have heard from two testers now that the share dialog does not show on their iPad, even though it does work on the same devices in the Simulator. You can see the code where I am calling `shareFile` at https://github.com/zenobase/zenobase-durations/blob/001cbe867b8eb8b47286c58f98b366602bc56dfb/lib/widgets.dart#L68.

pboos commented:
@ejain The only thing I see that could be the problem is that you are not using the `sharePositionOrigin` parameter like in the example:
https://github.com/pboos/plugins/blob/feature/shareFile/packages/share/example/lib/main.dart#L97

Maybe try using `sharePositionOrigin` as in the example and see if it solves the problem.

ejain commented:
When I specify a `sharePositionOrigin` with a zero offset as shown in the example, the resulting box dimensions are 768,1024, and only a few pixels of the share dialog are visible in the top left corner (6th gen iPad in the simulator). Using the PopupMenuButton's context instead, the share dialog again shows as expected in the simulator. Testing on an actual iPad TBD.

ejain commented:
Just confirmed that there is at least one iPad out there on which the share dialog works; the key to making it work on iPads was to set `sharePositionOrigin` to a specific element on the screen.

pboos commented:
@collinjackson anything else to do to get this pull request moving? current failing tests were only due to some http timeouts.

neckaros commented:
> Just confirmed that there is at least one iPad out there on which the share dialog works; the key to making it work on iPads was to set `sharePositionOrigin` to a specific element on the screen.

On the fix I did for esys share I added a check for width an height : https://github.com/esysberlin/esys-flutter-share/blob/26a2f34ddf70f7361c529677dad15aadd99193a6/ios/Classes/SwiftEsysFlutterSharePlugin.swift#L95 

If the context rect is the whole window then you have an issue as it will be equal to the bounds => no space for the share sheet

MaskyS commented:
@pboos https://github.com/aarajput/wc_flutter_share/issues/10 Would adding this fix finally complete this PR?

MaskyS commented:
@pboos, Tried sharing a PDF File with:
```
Share.shareFile(File(filePath),
        mimeType: 'application/pdf',
        text: "Text",
        subject: "Subject");
```

It works, but I got the following exception in my console:
```
E/DatabaseUtils(20041): Writing exception to parcel
E/DatabaseUtils(20041): java.lang.SecurityException: Permission Denial: reading androidx.core.content.FileProvider uri content://com.user.app.flutter.share_provider/external_cache/share/mypdf.pdf from pid=11289, uid=1000 requires the provider be exported, or grantUriPermission()
E/DatabaseUtils(20041): 	at android.content.ContentProvider.enforceReadPermissionInner(ContentProvider.java:729)
E/DatabaseUtils(20041): 	at android.content.ContentProvider$Transport.enforceReadPermission(ContentProvider.java:602)
E/DatabaseUtils(20041): 	at android.content.ContentProvider$Transport.query(ContentProvider.java:231)
E/DatabaseUtils(20041): 	at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:104)
E/DatabaseUtils(20041): 	at android.os.Binder.execTransactInternal(Binder.java:1021)
E/DatabaseUtils(20041): 	at android.os.Binder.execTransact(Binder.java:994)
```
This isn't related to the actual storage permission, as the exception is thrown irrespective of whether I've granted it.

My flutter doctor:

```
flutter doctor -v
[√] Flutter (Channel beta, v1.14.6, on Microsoft Windows [Version 10.0.18362.657], locale en-US)
    • Flutter version 1.14.6 at C:\src\flutter
    • Framework revision fabeb2a16f (7 weeks ago), 2020-01-28 07:56:51 -0800
    • Engine revision c4229bfbba
    • Dart version 2.8.0 (build 2.8.0-dev.5.0 fc3af737c7)

 
[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at C:\Users\kufah\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 28.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b04)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Android Studio (version 3.6)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 43.0.2
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b04)

[√] VS Code, 64-bit edition (version 1.42.1)
    • VS Code at C:\Program Files\Microsoft VS Code
    • Flutter extension version 3.8.1

[√] Connected device (3 available)
    • Pixel 2 XL • 712KPWQ1289633 • android-arm64  • Android 10 (API 29)
    • Chrome     • chrome         • web-javascript • Google Chrome 80.0.3987.132
    • Web Server • web-server     • web-javascript • Flutter Tools

• No issues found!
```

I looked it up on SO, and adding the following lines to `Share.java` gets rid of the error:

```
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
// the rest of the code
// ...
  void shareFiles(List<String> paths, List<String> mimeTypes, String text, String subject)
      throws IOException {
// ...
    List<ResolveInfo> resInfoList = activity.getPackageManager().queryIntentActivities(chooserIntent, PackageManager.MATCH_DEFAULT_ONLY);
    for (ResolveInfo resolveInfo : resInfoList) {
      String packageName = resolveInfo.activityInfo.packageName;
      activity.grantUriPermission(packageName, fileUris.get(0), Intent.FLAG_GRANT_WRITE_URI_PERMISSION | Intent.FLAG_GRANT_READ_URI_PERMISSION);
    }

    if (activity != null) {
      activity.startActivity(chooserIntent);
    } else {
      chooserIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
      activity.startActivity(chooserIntent);
    }
  }
// ...
```

Hope we can finally wrap this up after 4 years 😅😁

pboos commented:
@MaskyS thanks for the pull request on this pull request :D . I have merged it in.

I really hope this will get merged in soon. Has been open way too long.

JonasVautherin commented:
@collinjackson: May I ask how the PR reviews of external contributions are organized? Is there a priority list somewhere, or is there just not enough resources to review them at all?

aloisdeniel commented:
I also added support to save exported images to gallery to iOS gallery.

[PR #2](https://github.com/pboos/plugins/pull/2)

cms103 commented:
One challenge with this current implementation is the use of androidx.core.content.FileProvider as the Provider in the manifest.  This conflicts with other plugins (e.g. open_file) that use the same provider name.

Using a custom provider is fairly trivial, it was done for image_picker in https://github.com/flutter/plugins/pull/716

Draft PR here: https://github.com/pboos/plugins/pull/3/files

rafaelmpessoa commented:
please merge I'm waiting for it 😄

JonasVautherin commented:
@rafaelmpessoa: It doesn't seem like there is any interest from the flutter team, so you should use it as-is. @pboos wrote instructions in the beginning of this PR on how to use it directly.

cms103 commented:
@googlebot I consent.

aloisdeniel commented:
@googlebot I consent.

allanwolski commented:
What's missing for this to be merged?

JonasVautherin commented:
@allanwolski: I think interest/resources from maintainers. Looking at their github activity, it seems like @collinjackson and @amirh haven't been super active recently (https://github.com/flutter/plugins/graphs/contributors). Note that there seems to be activity in the main flutter repo: https://github.com/flutter/flutter/graphs/contributors.

Not sure what's happening with flutter/plugins, though. Maybe they are dropping support for that, or maybe they are not interested in community contributions. @ditman has been active recently, hopefully we will get news at some point :-).

Maybe the best way would be to give up contributing it to flutter/plugins, and just maintain it as a third-party plugin.

ditman commented:
I'm tracking who owns this plugin currently to have this PR merged. Can anybody (@pboos?) update the branch so there's no conflicts with master? I also couldn't find any tests for the Java side, I'm not sure if those are a thing (I normally work on web plugins ;))

pboos commented:
Thx @ditman for tracking this! I updated the branch and resolved the conflicts. I tested it and the example still works fine. 

There were no tests before this update, so I guess it should be fine without adding any in this PR 😄.

ditman commented:
// @cyanglaz this is the PR we talked about a few days ago!

ditman commented on packages/share/lib/share.dart:
```diff
@@ -51,4 +52,128 @@ class Share {
+    return channel.invokeMethod('shareFiles', params);
+  }
+
+  static String _mimeTypeForFile(File file) {
```
How about using: 

https://pub.dev/packages/mime

That way you don't have to maintain this list yourself.

pboos replied:
Good point. Switching to it now.

ditman commented on packages/share/lib/share.dart:
```diff
@@ -51,4 +52,128 @@ class Share {
+      case 'txt':
+        return 'text/plain';
+    }
+    return null;
```
Returning null when unknown seems to go against the spec, [MDN says](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types#Discrete_types):

> Generic binary data (or binary data whose true type is unknown) is `application/octet-stream`.

(But if you replace this by the mime library they probably handle that themselves)

pboos replied:
Checked what happens with `mime` package and that one returns null. So adjusting my code to return `application/octet-stream` in case `mime` returns null.

ditman commented on packages/share/lib/share.dart:
```diff
@@ -3,6 +3,7 @@
 // found in the LICENSE file.
 
 import 'dart:async';
+import 'dart:io';
```
Thinking of a future version for web here. Do we need to share `dart:io` File objects? Can we tell the native code to share "file paths" or similar, and have the file be created in the native side?

Depending on `dart:io` File objects guarantees that we'll have to rewrite this if we want to add Web support (even though [the API](https://caniuse.com/#feat=web-share) is not very well supported yet)

pboos replied:
Makes sense. I have quickly changed it and confirmed it still works. Only difference now is that if a non existent path is passed in, the native side will throw an error. But I guess that is okay and the expected result. Before it threw an assertion error from the dart side.

ditman commented on packages/share/lib/share.dart:
```diff
@@ -51,4 +52,128 @@ class Share {
+    assert(files.isNotEmpty);
+    assert(files.every((file) => file.existsSync()));
+    final Map<String, dynamic> params = <String, dynamic>{
+      'paths': files.map((file) => file.path).toList(),
```
Ah, you end up converting this to paths to pass them through the platform interface.

Can you make the API of this method accept `List<String> paths`? That'd make this a little bit more portable!

pboos replied:
Done

pboos commented:
@ditman Should I already update the version to `0.6.5` and also update the `CHANGELOG.md` file? Or does someone else normally do that?

ditman commented:
@pboos please, do update the version and CHANGELOG.md, otherwise the `analyze` test won't pass :)

pboos commented:
@ditman done. Currently some checks are failing due to `integration_test` it seems. Or do you have any idea how I can resolve that? Did not seem to occur before I switched from `e2e` to `integration_test`.

Nimrodda commented on packages/share/android/src/main/AndroidManifest.xml:
```diff
@@ -1,3 +1,14 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
   package="io.flutter.plugins.share">
+    <application>
+      <provider
+        android:name="io.flutter.plugins.share.ShareFileProvider"
+        android:authorities="${applicationId}.flutter.share_provider"
+        android:exported="false"
+        android:grantUriPermissions="true">
+        <meta-data
+          android:name="android.support.FILE_PROVIDER_PATHS"
+          android:resource="@xml/flutter_share_file_paths"/>
+      </provider>
```
I was wondering [merge rule markers](https://developer.android.com/studio/build/manifest-merge#merge_rule_markers) can be of use here instead of introducing another provider 🤔

pboos replied:
Good point. This was as well what I hated most in this plugin. But wonder what other provider to merge it with and if meta-data could make problems if merged with another. Specially if the other package as well uses FILE_PROVIDER_PATHS with a different config. Basically I know too little about how to handle this here. If anyone has more knowledge and can tell me how (or even create a PR on to update my feature branch) that would be great :)

ditman commented:
Hmmm @pboos, I think the current master on this branch might have some breakage (not your bad, btw), I'll let you know when flutter/plugins:master is stable for you to attempt another rebase.

_(That should fix the build_all tests.)_

ditman commented:
Master is green now, at least until this point: https://github.com/flutter/plugins/commit/e0a49a7fbdbc4e06d04e4cfbbe2c6056f09bf3d4

@pboos if you rebase your branch again, we can see if the failures come from your code, or were some staleness in other parts of the code.

Thanks for your patience!

pboos commented:
@ditman thank you for the heads up. Seems like it is all green now except for the master branch failing (due to timeouts?).

ditman commented on packages/share/example/lib/main.dart:
```diff
@@ -69,11 +91,30 @@ class DemoAppState extends State<DemoApp> {
+
+                              if (imagePaths.isNotEmpty) {
+                                if (imagePaths.length == 1) {
+                                  await Share.shareFile('',
```
Shouldn't this share `imagePaths.first` instead of an empty string?

pboos replied:
Not sure why I never noticed this. But removing this anyway in favour of `shareFiles()`

ditman commented on packages/share/example/lib/main.dart:
```diff
@@ -83,4 +124,51 @@ class DemoAppState extends State<DemoApp> {
     );
   }
+
+  Widget _buildImagePreviews() {
```
Why not extract these functions (buildImagePreviews and buildImagePreview) to stateless widgets in a separate image_preview.dart file?

pboos replied:
Good point, makes it more readable. Done so now :)

ditman commented on packages/share/lib/share.dart:
```diff
@@ -51,4 +52,79 @@ class Share {
+  ///
+  /// May throw [PlatformException] or [FormatException]
+  /// from [MethodChannel].
+  static Future<void> shareFile(
```
I'm not sure the `shareFile` provides a lot of value. Why not remove it in favor of only one single `shareFiles` method that accepts an array?

(Also look at the example, it makes the code a little bit more complex having to decide whether to call `shareFile` vs `shareFiles`

JonasVautherin replied:
I think `shareFile()` was there first, and `shareFiles()` was contributed later. That may explain its existence. I do agree that `shareFiles()` works with an array of one file, too :+1:

pboos replied:
Yes I think `shareFile()` was there first. I left it for making things easier when just sharing one file. But I agree that it just removes `[]` which is not enough to keep it.

pboos commented:
@ditman did the corrections you suggest. I agree the `shareFile` did not really add any value. And seems like all tests passed now as well. I hope we can get somebody to review Android/iOS. Looks like we are getting closer to getting this merged 🤞 .

cyanglaz commented on packages/share/ios/Classes/FLTSharePlugin.m:
```diff
@@ -69,19 +121,37 @@ + (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
+      for (NSString *path in paths) {
+        if (path.length == 0) {
+          result([FlutterError errorWithCode:@"error"
+                                     message:@"Non-empty paths expected"
```
Maybe `Each path in `paths` must not be empty`?

pboos replied:
Copy paste mistake 🙈 . Replaced message with `Each path must not be empty`. Felt like `each path in paths` was not adding much more info.

cyanglaz commented on packages/share/ios/Classes/FLTSharePlugin.m:
```diff
@@ -91,18 +161,56 @@ + (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
+          atSource:(CGRect)origin {
+  NSMutableArray *items = [[NSMutableArray alloc] init];
+
+  if (text != nil || subject != nil) {
```
nits: can be shorten as `if (!text || !subject)`

pboos replied:
I guess you meant `if (text || subject)` :)

cyanglaz commented:
@pboos Thanks for the PR! It looks good overall. I left some comments around java and objc implementations.

pboos commented on packages/share/android/src/main/java/io/flutter/plugins/share/Share.java:
```diff
@@ -47,4 +61,150 @@ void share(String text, String subject) {
+    ArrayList<Uri> fileUris = getUrisForPaths(paths);
+
+    Intent shareIntent = new Intent();
+    if (fileUris.size() == 1) {
```
Should never be 0 because the dart side handles it. But just to be sure I now added an `isEmpty` check that then just does a normal share with `share(text, subject)`.

pboos commented on packages/share/android/src/main/java/io/flutter/plugins/share/Share.java:
```diff
@@ -47,4 +61,150 @@ void share(String text, String subject) {
+      activity.startActivity(chooserIntent);
+    } else {
+      chooserIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+      activity.startActivity(chooserIntent);
```
Good point :). Seems like this was code that I copied over from master. The already existing `share()` method has the same. But agree this would crash. I changed the code now to handle this better and even just use the application context if no activity is attached.

pboos commented on packages/share/example/pubspec.yaml:
```diff
@@ -20,4 +21,3 @@ flutter:
 environment:
   sdk: ">=2.0.0-dev.28.0 <3.0.0"
   flutter: ">=1.9.1+hotfix.2 <2.0.0"
-
```
There is an empty line (line 22 ends with a line break). I thought two empty lines was too much. I think github does not not show the empty line.

ruskakimov replied:
@pboos is right, there is no "No newline at end of file" warning.
![image](https://user-images.githubusercontent.com/5536948/91537785-6edbbb80-e949-11ea-9535-fb6c54e3684e.png)

cyanglaz replied:
SG. Thanks!

pboos commented on packages/share/ios/Classes/FLTSharePlugin.m:
```diff
@@ -57,8 +98,19 @@ + (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
+    NSNumber *originHeight = arguments[@"originHeight"];
+
+    CGRect originRect = CGRectZero;
+    if (originX != nil && originY != nil && originWidth != nil && originHeight != nil) {
```
Nice :). Just to ensure, objective-c is not like JavaScript in this place? in Javascript `originX = 0` would fail the `if (originX)` check as 0 resolves to false. I hope in objective-c that is not he case?

ditman replied:
It seems obj-C behaves the same as Javascript: https://stackoverflow.com/a/6177766

ditman replied:
But in this case, it seems `originWhatevers` are Strings that get [cast to a double](https://developer.apple.com/documentation/foundation/nsstring/1414031-doublevalue?language=objc) on line 109?

`[originX doubleValue]`...

ditman replied:
[NSNumber seems to cast to false when it's `0`](https://developer.apple.com/documentation/foundation/nsnumber), I'd leave the `!= nil` check in this case too :P

pboos replied:
Thx for checking this :). I was not sure either and could not find a clear answer. Tested around a bit and think we are save without the nil check.

pboos commented on packages/share/android/src/main/java/io/flutter/plugins/share/Share.java:
```diff
@@ -47,4 +61,150 @@ void share(String text, String subject) {
+    Intent chooserIntent = Intent.createChooser(shareIntent, null /* dialog title optional */);
+
+    List<ResolveInfo> resInfoList =
+        activity
```
Yes, this could be problematic. Found some other places as well and fixed them now.

pboos commented:
@cyanglaz Thanks a lot for the great comments. I acted on them and now the code should be a lot better :). You caught some embarrassing bugs that snuck their way in there :)

ditman commented:
All looks green, landing this now, will publish ASAP! Thanks for the contribution (and the patience) @pboos!!

ditman commented:
`share: ^0.6.5` has been tagged and [published](https://pub.dev/packages/share/changelog#065)!

lazylazyllama commented:
Now as we have to put in the path of our files I wonder how I can share a base64 encoded File that I get from the network?

pboos commented:
@ditman Thanks a lot for taking this on and getting it merged! 🚀 

@lazylazyllama You could always temporarily store that file on the device within the app. The user won't even notice it. And as a note, before it was `File`, now it is the path. But even before it just forwarded the absolute path of the file in the background. So if something worked before, you can just do `file.path` now, as that is what it did before the recent changes on this PR.
